components/support/sql/src/open_database.rs

```diff
@@ -87,21 +89,70 @@ pub fn open_database_with_flags<CI: ConnectionInitializer, P: AsRef<Path>>(
     path: P,
     open_flags: OpenFlags,
     connection_initializer: &CI,
+) -> Result<Connection> {
+    let oi = OpenInfo {
+        path: path.as_ref(),
+        open_flags,
+        connection_initializer,
+        handle_fatal_errors: true,
+    };
+    do_open_database_with_flags(oi)
+}
+
+// This structure helps us with our error recovery strategy.
+struct OpenInfo<'a, CI: ConnectionInitializer> {
+    path: &'a Path,
+    open_flags: OpenFlags,
+    connection_initializer: &'a CI,
+    handle_fatal_errors: bool,
+}
+
+fn do_open_database_with_flags<CI: ConnectionInitializer>(
+    open_info: OpenInfo<'_, CI>,
 ) -> Result<Connection> {
     // Try running the migration logic with an existing file
     log::debug!("{}: opening database", CI::NAME);
-    let mut conn = Connection::open_with_flags(path, open_flags)?;
-    let run_init = should_init(&conn)?;
+    // It turns out Connection::open_with_flags appears to always return without checking
+    // the validity - but that might not be true in all cases, so be safe.
+    let mut conn = match Connection::open_with_flags(open_info.path, open_info.open_flags) {
```

Comment:
everywhere that can fail needs to do this kind of dance, which is a bit ugly, but I can't see a better way.

Reply:
FWIW I don't think it looks too bad, the only thing I can think of is maybe a macro `open_with_flags!` that abstracts but probably won't make it that much cleaner